pkg/activator/revision_backends.go

```diff
@@ -0,0 +1,285 @@
+// information for how to reach a revision via PodIP
+type RevisionDestsUpdate struct {
+	Rev   RevisionID
+	Dests []string
+}
+
+const (
+	probePortName  string        = "http"
+	probeTimeout   time.Duration = 300 * time.Millisecond
+	probeFrequency time.Duration = 200 * time.Millisecond
+)
+
+type revisionWatcher struct {
+	rev          RevisionID
+	updateCh     chan<- *RevisionDestsUpdate
+	dests        []string
+	healthStates map[string]bool
+
+	transport http.RoundTripper
+	destsChan <-chan []string
+	logger    *zap.SugaredLogger
+}
+
+func newRevisionWatcher(rev RevisionID, updateCh chan<- *RevisionDestsUpdate,
+	destsChan <-chan []string, transport http.RoundTripper,
+	logger *zap.SugaredLogger) *revisionWatcher {
+	return &revisionWatcher{
+		rev:          rev,
+		updateCh:     updateCh,
+		healthStates: make(map[string]bool),
+		transport:    transport,
+		destsChan:    destsChan,
+		logger:       logger,
+	}
+}
+
+func endpointsToDests(endpoints *corev1.Endpoints) []string {
+	ret := []string{}
+
+	for _, es := range endpoints.Subsets {
+		portVal := int32(0)
+		for _, port := range es.Ports {
+			if port.Name == probePortName {
+				portVal = port.Port
+				break
+			}
+		}
+
+		// This endpoint doesnt have a port we want to probe
+		if portVal == 0 {
+			continue
+		}
+
+		for _, addr := range es.Addresses {
+			// Prefer IP as we can avoid a DNS lookup this way
+			ret = append(ret, net.JoinHostPort(addr.IP, strconv.Itoa(int(portVal))))
+		}
+	}
+
+	return ret
+}
+
+func filterHealthyDests(dests map[string]bool) []string {
+	ret := make([]string, len(dests))
+	i := 0
+	for dest, healthy := range dests {
+		if healthy {
+			ret[i] = dest
+			i++
+		}
+	}
+	return ret[:i]
+}
+
+type destHealth struct {
+	dest   string
+	health bool
+}
+
+func (rw *revisionWatcher) checkDests() {
+	healthStatesCh := make(chan *destHealth, len(rw.dests))
+	defer close(healthStatesCh)
+
+	// Context used for our requests
+	ctx, cancel := context.WithTimeout(context.Background(), probeTimeout)
+	defer cancel()
+
+	for _, dest := range rw.dests {
+		// If the dest is already healthy then save this
+		if curHealthy, ok := rw.healthStates[dest]; ok && curHealthy {
+			healthStatesCh <- &destHealth{dest, true}
+			continue
+		}
+
+		go func(dest string) {
+			httpDest := url.URL{
+				Scheme: "http",
+				Host:   dest,
+			}
+			ok, err := prober.Do(ctx, rw.transport, httpDest.String(),
+				prober.WithHeader(network.ProbeHeaderName, queue.Name),
+				prober.ExpectsBody(queue.Name))
+
+			healthStatesCh <- &destHealth{dest, ok}
+
+			if err != nil {
+				rw.logger.Errorw("Failed probing "+dest, zap.Error(err))
+			} else if !ok {
+				rw.logger.Info("Probing of dest " + dest)
+			}
+		}(dest)
+	}
+
+	healthStates := make(map[string]bool, len(rw.dests))
+	for i := 0; i < len(rw.dests); i++ {
+		dh := <-healthStatesCh
+		healthStates[dh.dest] = dh.health
+	}
+
+	if !reflect.DeepEqual(rw.healthStates, healthStates) {
```

Comment:
Do we have any concerns with the performance of this call? We can check in the loop above this if we apply any changes and elide this I think. Using `reflect` makes me slightly uncomfortable, but that might be my JVM muscle-memory.

Reply:
I dont think we can easily check this above - that can check if we modified any of the current dests but it cant easily detect e.g. a dest was removed. I did some research and the reflect call does the optimizations we'd want (checks len first and short circuits if it can) so I think its the best we can do.